man/nbkit.1
.\" Process this file with "groff -man -Tascii foo.1" or "nbman2ht foo.1"
.\" See NodeBrain Developer's Guide for project standards
.\"
.TH nbkit 1 "December 2014" "NodeBrain 0.9.03" "NodeBrain Administrator Guide"
.SH NAME
\fBnbkit\fP - NodeBrain Kit Utility
.SH SYNOPSIS
.B nbkit
{ \fB-c\fP | \fB--caboodles\fP | \fB-k\fP | \fB--kits\fP }

.P
\fBnbkit\fP \fIcaboodle\fP \fIverb\fP [\fIarguments\fP]

.P
\fBnbkit\fP \fIcaboodle\fP \fBlink\fP \fIdirectory\fP

.P
\fBnbkit\fP \fIcaboodle\fP \fBunlink\fP

.P
\fBnbkit\fP \fIcaboodle\fP \fBuse\fP \fIkit\fP

.P
\fBnbkit\fP \fIcaboodle\fP { \fBexport\fP | \fBimport\fP | \fBupgrade\fP } \fIarchive-file\fP

.P
.B nbkit
\fIcaboodle\fP 
.B alarm
[ \fBroute="\fP...\fB"\fP ]
[ \fBappl="\fP...\fB"\fP ]
[ \fBgroup="\fP...\fB"\fP ]
[ \fBnode="\fP...\fB"\fP ]
[ \fBseverity="\fP...\fB"\fP ]
[ \fBtext="\fP...\fB"\fP ]

.P
.B nbkit
\fIcaboodle\fP {
.B archive
|
.B bounce
|
.B config
|
.B connect
|
.B start
|
.B stop
} \fIagent\fP

.P
.B nbkit
\fIcaboodle\fP {
.B commit
|
.B compile
|
.B disable
|
.B edit
|
.B enable
|
.B list
|
.B setup
|
.B view
} \fIplan\fP

.SH DESCRIPTION
The \fBnbkit\fP(1) command is used to manage
NodeBrain applications that conform to the \fBNodeBrain Kit and Caboodle Framework\fP
described by the \fBnbkit.caboodle\fP(7) manual page.
While the \fBnbkit\fP(1) command is provided with the NodeBrain rule engine, it is designed to
work with caboodles built using NodeBrain kits.  
Caboodles reside on systems where the rule engine is installed,
while kits are development tools not normally installed on production systems.
 
The \fBCaboodle NodeBrain Kit\fP provides the key components of a caboodle necessary 
for many of the nbkit command options described here to work.  Without a caboodle constructed from the
Caboodle NodeBrain Kit (or a compatible package), the nbkit command has little or no value.
The Caboodle NodeBrain Kit also provides caboodle components enabling a web interface for
browser access to most of the same operations performed using the nbkit command.

The nbkit command is primarily a method of invoking an nbkit_ command within a caboodle.  This means
the actual behavior of many of the commands described here will depend of the version of the Caboodle
NodeBrain Kit used to create your caboodle.

The following common terms have a specific meaning within the context of a NodeBrain kit.
.IP \fBcaboodle\fP 11
A directory containing a collection of components organized into
a predefined set of subdirectories.  Components make reference to other components using paths
relative to the caboodle.  This makes a caboodle portable.

.IP \fBkit\fP
A collection of components organized like a caboodle, but intended
for distribution and used to construct caboodles or add functionality to existing caboodles.
A caboodle may be composed from multiple kits.

.IP \fBagent\fP
A NodeBrain script that runs as a background process, or daemon. 

.IP \fBplan\fP
An XML document used to represent a set of NodeBrain rules or configuration files based on
a model.  A plan compiler converts a plan from XML document format into a target format, often
NodeBrain rules.
.P

.SH ARGUMENTS

.IP \fB-c\fP 17
Display a list of defined caboodles.
.IP \fB--caboodle\fP

.IP \fB-k\fP
Display a list of installed NodeBrain kits.
.IP \fB--kits\fP

.IP \fIcaboodle\fP
This argument is specified on most nbkit commands to identify the caboodle upon which to operate.
The value specified is a single word alias for the caboodle path.
See the \fBlink\fP command in the next section.

.IP \fIverb\fP
This argument identifies the operation to be performed on the specified caboodle.

.IP \fIarguments\fP
Additional arguments depend on the verb specified.
.P

.SH CABOODLE COMMANDS
Commands that operate on a caboodle without reference to specific components are described in this section.
Commands that operate on agents and plans are described in the next two sections. 
.IP \fBlink\fP 11
Use to define a caboodle.  This command simply creates a symbolic link in the home
directory of the user under a subdirectory of ".nb/caboodle".
Other commands make use of the symbolic link to set the working directory before performing an operation.

    nbkit \fIcaboodle\fP link \fIdirectory\fP

For a caboodle directory "/var/foobar" that you want to reference as "foobar",
you would enter the following command.

    nbkit foobar link /var/foobar

.IP \fBunlink\fP
Use to undefine a caboodle. 

    nbkit \fIcaboodle\fP unlink

If you want to redefine a caboodle to use a different directory, it is not necessary to use "unlink".
You may simply issue a "link" command again specifying a different directory.

.IP \fBuse\fP
The "use" verb associates a caboodle with an installed kit.  This makes the components
provided by the kit available to the caboodle.

    nbkit \fIcaboodle\fP use \fIkit-directory\fP

To associate nbkit-system-0.8.14 with a caboodle, the command would be as follows.

    nbkit \fIcaboodle\fP use /usr/share/nbkit/nbkit-system-0.8.14

.IP \fBexport\fP
Use to export a caboodle to a compressed tar file as a backup, or to copy to another server.
Log files and other files considered to be caboodle specific and of no value to similar caboodles are not exported.

    nbkit \fIcaboodle\fP export \fIarchive-file\fP

For example:

    nbkit mycaboodle export /tmp/mycaboodle.tar.gz


.IP \fBimport\fP
Use to import an exported caboodle into a similar caboodle without replacing locally modified components, and without
disturbing components in the caboodle not found in the archive.
This is appropriate in cases where there is local maintenance performed downstream from the source of the archive, including
cases where a caboodle is a combination of imports from multiple upstream archive providers.

    nbkit \fIcaboodle\fP import \fIarchive-file\fP

For example:

    nbkit mycaboodle2 import /tmp/mycaboodle.tar.gz

.IP \fBupgrade\fP
Use to upgrade a caboodle from an exported caboodle, without protecting local modifications and additions.
This is appropriate when following a policy of not making local modifications, and the archive provides
a complete set of application components.

    nbkit \fIcaboodle\fP upgrade \fIarchive-file\fP

For example:

    nbkit mycaboodle2 upgrade /tmp/mycaboodle.tar.gz

.IP \fBalarm\fP
Generate an alarm to subscribers.
This command sends a NodeBrain alert command to the alarm node of the Caboodle agent supporting the specified
caboodle.  Multiple optional alarm attributes may be specified.  Some have default values when not specified.
All alarm attributes may be referenced by an alarm subscription, which is required for actual distribution of the
alarm.

    nbkit \fIcaboodle\fP alarm \fIattribute\fP="\fIvalue\fP" [ ... ]

Specify one or more alarm attributes:

.RS 
.IP \fBappl\fP 11
Specify the name of the application to which the alarm applies.

.IP \fBgroup\fP
Specify a value that categories the alarm; e.g. "OS", "Database", "Web".  The group
attribute is intented to route alarms to a unique group of subject experts.  This is unlike the route
attributes which may be used to provide visibility to various interested recipients.

.IP \fBnode\fP
Specify the hostname associated with the alarm.  By default, the local hostname is used.

.IP \fBroute\fP
Specify a comma seperated list of routing codes for alarm subscriptions.
The codes are strings that have meaning within a given application.  For example, if the caboodle is a
security event monitoring application, a value of "SOC" might be used to indicate the alarm is intented for
the Security Operations Center.  Multiple codes may be entered; e.g. "SOC,NOC".

.IP \fBseverity\fP
Specify a level of severity for the alarm.  Typical values are "critical", "major", "minor", "warning" and "info",
although you may select alternate values that work better in the context of your application.

.IP \fBtext\fP
Specify message text that describes the alarm condition; e.g. "Roof flew off the barn".  You may include a message
identifier of the form AAANNNN at the start of the text; e.g. "PWS0125 Roof flew off the barn".
.RE
.P

.SH AGENT COMMANDS
Commands in this section operate on NodeBrain agents and are handed off
to the bin/nbkit_agent command within the specified caboodle.
An \fIagent\fP is a NodeBrain script that runs as a daemon.

    nbkit \fIcaboodle\fP \fIverb\fP \fIagent\fP

The \fIagent\fP parameter must specify the name of an agent plan and/or the name of a NodeBrain script in the agent
subdirectory; e.g. "Caboodle".
To start the Caboodle agent provided by the Caboodle NodeBrain Kit, the command would be:

    nbkit \fIcaboodle\fP start Caboodle

Specify one of the following verbs:
.IP \fBarchive\fP 11
Start a new log file after renaming the current log file with a current time stamp.

.IP \fBbounce\fP
Stop and start the agent.  If the current expected state is "down", the agent will be started and the expected
state will change to "up".

.IP \fBcheck\fP
Check expected state of an agent and change the current state to match.
If an agent is expected to be "up" but is currently "down", it is started.
If it is "up" and expected "down", it is stopped.

.IP \fBconnect\fP
Invoke the NodeBrain interpreter in interactive mode and prompt for commands to be issued to the specified agent.

.IP \fBstart\fP
Start the specified agent if it is currently "down" and set the expected state to "up".

.IP \fBstop\fP
Stop the specified agent if it is currently "up" and set the expected state to "down".
.P

.SH PLAN COMMANDS
Commands in this section operate on NodeBrain plans and are handled by
the bin/nbkit_plan command within the specified caboodle.
A \fIplan\fP is an XML document that normally translates into a set of NodeBrain rules
or a configuration file based on a model.  Compilers are provided to perform this
translation.  When an agent is managed as a plan, just one type of plan,
these commands may be used to operate on the agent. 

    nbkit \fIcaboodle\fP \fIverb\fP \fIplan\fP

Operations performed by these commands may also be performed by the NodeBrain Planner
web interface.  

.IP \fBcompile\fP 11
Convert a plan XML document into the target format.
.IP \fBdisable\fP
Put a plan in a disabled state, or disable a relationship between two plans.
.IP \fBedit\fP
Edit a plan configuration file.
The XML document is converted into a temporary file with a less complex format and vi is invoked 
to edit the file.
After exiting vi, the file is converted back into XML document format.
The XML document is then optionally converted into a target format by the associated compiler.

.IP \fBenable\fP
Allow a plan to generate alerts and start probe.
.IP \fBhelp\fP
Display this list of commands.
.IP \fBlist\fP
Display a list of configured plan.
.IP \fBremove\fP
Remove a plan from configuration.
.IP \fBrename\fP
Rename a plan.
.IP \fBsetup\fP
Invoke assisted setup script for a plan.
.IP \fBshow\fP
Display plan relationships.
.IP \fBstatus\fP
Display current status of plans.
.IP \fBview\fP
Browse a plan configuration file.
.P

.SH FILES
.IP ~/.nb/\fIcaboodle\fP
The caboodles are defined by a user, the caboodle name is defined as a symbolic
link to the caboodle directory.  The links are located in the ".nb/caboodle" subdirectory
of the home directory of the user.
.IP /usr/share/nbkit/\fIkit\fP
When kits are installed, the platform indepenent components are in this directory
or "/usr/local/share/nbkit/\fIkit\fP
.P

.SH HISTORY
The nbkit command was developed as part of the Caboodle NodeBrain Kit
in 2006.  It derives from a similar scheme developed for
NodeBrain in 1998 called the System Monitor Kit or Sysmon.

.SH BUGS
Please report bugs to <bugs@nodebrain.org> or enter a bug report
at http://nodebrain.org.

.SH AUTHOR
Ed Trettevik <eat@nodebrain.org>

.SH DOCUMENTATION
Online documentation is available at http://nodebrain.org.
Manuals are also available in Texinfo format in the
git repository.  These documents are included in source
distribution files.

.SH SEE ALSO
.\"ht page
\fBnbkit.caboodle\fP(7)
.\"ht page
\fBnodebrain\fP(7),
.\"ht page
\fBnb\fP(1)


NEWS
 _   _           _      ____
| \ | |         | |    |    \          ( )
|  \| | ___   __| | ___| | ) |_ _  __ _ _ _ __
|     |/   \ /    |/ _ \    <|  _|/    | |  _ \
| |\  | ( ) | ( | |  __/ | ) | |   ( | | | | | |
|_| \_|\___/ \__ _|\___|____/|_|  \__ _|_|_| |_|

============================================================

NodeBrain 0.9.03 (Columbo) Change Summary

* Released December 14, 2014

* Theme:  Reduced Sensitivity to Command Arrival Order

     This version provides more flexibility in referencing
     nodes before they are defined. It also automatically
     resets event attributes to the unknown state when not
     specified on an alert, eliminating the need to specify
     all event attributes on every alert---required under
     prior versions to avoid event attribute values from
     impacting rule response to a subsequent alert.

* Packaging:

  o  The version scheme for the NodeBrain Library (libnb)
     has switched to using the Application Binary Interface
     version, instead of the release version based scheme
     used previously.  The release version approach was
     intended to require applications to build for every
     release, assuming there might be incompatibilities in
     every release. The change to an ABI based version is
     being made in preparation for a 1.0 release and 
     subsequent 1.x releases that will not require using
     applications to rebuild until release 2.0, except to
     gain access to enhancements in a 1.x release.

        libnb-0.9.02.so    [0.9.02 prior release]
        libnb.so.0.0.903   [0.9.03 this release]
        libnb.so.1.0.1000  [1.0.00 future release]

  o  The rpm spec file, and debian packaging files have 
     been modified to get better build results on various
     platforms after the switch from readline to libedit.
     The effort to make the project friendly to package
     maintainers for various Linux distributions is still
     in progress. However, rpmlint is silent on CentOS so
     the package will be submit to Fedora for review.

* Documentation:

  o  The NodeBrain Language manual has been updated to 
     cover new topics:

       - Alert attribute transience
       - Node functions
       - Cell functions
       - Redefine command (new in release 0.9.02)

     These topics were not covered previously because their
     introduction in the prior releases was experimental.
     Because of this, the incompatibilities introduced by
     this release for these features are expected to have
     little impact.

  o  The NodeBrain Library manual has been updated with
     additional API functions for:

       - Access to all terms defined within a node
         glossary.
       - Registration of simple cell functions.
     
* Incompatibilities:

  o  Applications using ALERT commands for both event
     and state monitoring may break under this release,
     requiring rule changes before upgrading.  You will
     only have a problem if you depend on some terms
     defined within a given node that are set by an ALERT
     to the node, retaining their value when subsequent
     ALERT commands to the same node do not set a value
     for the term.  This would mean some of the terms set
     by an ALERT are intended to express "state", while
     others express an "event".  Starting with this
     release, you will need to move all terms used to
     express "state" within ALERT commands to a different
     node. Otherwise, ALERT commands not setting the "state"
     terms will automatically cause them to revert to Unknown.
     This is not a problem for "state" terms that are set by
     all ALERT commands to the node. 

       define n node;
       n. alert a=1,b=2; # a and b may be state or event terms
       n. alert a=2,b=3; # No problem because same terms set
       n. alert a=3; # Problem if b is intended for "state"
       # Prior releases would leave b set to 3
       # This release resets b to Unknown

     If this is a problem for you, it can be resolved as
     follows.
       
       define b cell; # Define state term in different node
       define n node;
       n. alert a=1,b=2; # a is for "event", b is for "state"
       n. alert a=3; # b is not reset to Unknown

  o  The special syntax for alert attributes has been
     dropped after introduction in the prior release. Terms
     starting with "@" are now interpreted as node functions
     instead of transient alert attributes.  We are making
     this change now under the assumption that there has
     been no significant adoption of either alert attribute
     transient terms or node functions.

  o  The syntax for node functions has changed abruptly under
     the assumption that there has been no significant
     adoption of this feature after introduction in the
     prior release. Node functions were briefly called "facets"
     because they are named after the facet of a node that they
     manage or access.  But it seems more natural to call them
     functions that access facets of a node.

       Was:      node_facet
       Now:      node@function  -or-   @function

     This freed up "_" for navigation of multiple level
     glossaries of terms within a node, and enables function
     reference within the context node without repeating
     the node term.

* New Features:

  o  The ALERT command now automatically resets terms defined
     directly within the context node to Unknown if set by 
     the prior ALERT to the node and not the present ALERT.
     This eliminates the need to set all event attributes in
     every alert to avoid unintended values from prior alerts.
 
  o  Node functions can now be referenced within the context
     node without specifying the node term.

       node@function(arglist) # previously node_facet(argList)
       @function(argList)     # previously unsupported

  o  Node functions can now be referenced before a node is 
     associated with a skill that provides the function.  This

trunk/lib/nbkit.c
/*
* Copyright (C) 2014 Ed Trettevik <eat@nodebrain.org>
*
* NodeBrain is free software; you can modify and/or redistribute it under the
* terms of either the MIT License (Expat) or the following NodeBrain License.
*
* Permission to use and redistribute with or without fee, in source and binary
* forms, with or without modification, is granted free of charge to any person
* obtaining a copy of this software, provided that the above copyright notice,
* this permission notice, and the following disclaimer are retained with source
* code and reproduced in documentation included with binary distributions. 
*
* Unless required by applicable law or agreed to in writing, this software is
* distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
* KIND, either express or implied.
*
*==============================================================================
* Program:  NodeBrain
*
* File:     nbkit.c 
*
* Title:    nbkit Command Processor
*
* Function:
*
*   This file provides minimal nbkit command functionality and passes the rest
*   off to the nbkit_cmd script in the specified caboodle.
*
* Synopsis:
*
*   int nbKit(int argc,char *argv[]);
* 
* Description
*
*=============================================================================
* Change History:
*
*    Date    Name/Change
* ---------- -----------------------------------------------------------------
* 2013-02-03 Ed Trettevik (original prototype version introduced in 0.8.14)
*=============================================================================
*/
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <dirent.h>
#include <pwd.h>
#include <errno.h>
#include <unistd.h>

static int nbKitUsage(void){
    printf("Usage:\n");
    printf("  nbkit { -k | --kits }              : Display kits\n");
    printf("  nbkit { -c | --caboodles }         : Display caboodles\n");
    printf("  nbkit <caboodle> unlink            : Remove caboodle link\n");
    printf("  nbkit <caboodle> link <directory>  : Create caboodle link\n");
    printf("  nbkit <caboodle> <verb> ...        : Caboodle nbkit_ operations\n");
    printf("  nbkit <caboodle> help              : Display nbkit_ help\n");
    return(0);
    }

static int nbKitKits(void){
  DIR *dir;
  struct dirent *ent;
  int i;
  char *kitroot[2]={"/usr/share/nbkit","/usr/local/share/nbkit"};
  for(i=0;i<2;i++){
    dir=opendir(kitroot[i]);
    if(dir){
      while((ent=readdir(dir))!=NULL){
        if(*ent->d_name!='.'){
          printf("%s/%s\n",kitroot[i],ent->d_name);
          }
        }
      closedir(dir);
      }
    }
  return(0);
  }

static int nbKitCaboodles(char *home){
  DIR *dir;
  struct dirent *ent;
  char *subdir[2]={".nb",".nbkit"};
  char dirname[1024];
  char linkname[1024];
  char cabdir[1024];
  int i;
  ssize_t linksize;

  for(i=0;i<2;i++){
    snprintf(dirname,sizeof(dirname),"%s/%s/caboodle",home,subdir[i]); 
    dir=opendir(dirname);
    if(dir){
      fprintf(stderr,"Caboodles defined at %s\n",dirname);
      while((ent=readdir(dir))!=NULL){
        if(*ent->d_name!='.'){
          snprintf(linkname,sizeof(linkname),"%s/%s",dirname,ent->d_name);
          linksize=readlink(linkname,cabdir,sizeof(cabdir));
          if(linksize>0 && linksize<sizeof(cabdir)){
            *(cabdir+linksize)=0;
            printf("  %s -> %s\n",ent->d_name,cabdir);
            }
          else if(linksize<0){
            fprintf(stderr,"Error reading link %s - %s\n",linkname,strerror(errno)); 
            closedir(dir);
            return(1);
            }
          }
        }
      closedir(dir);
      }
    }
  return(0);
  }

static int nbKitUnlink(char *home,char *caboodle){
  char pathname[1024];
  int len,rc;
  
  len=snprintf(pathname,sizeof(pathname),"%s/.nb/caboodle/%s",home,caboodle);
  if(len<0 || len>=sizeof(pathname)){
    fprintf(stderr,"Unable to unlink %s - link path too long for buffer\n",caboodle);
    return(1);
    }
  rc=unlink(pathname);
  if(rc<0){
    fprintf(stderr,"Unable to unlink %s - %s\n",pathname,strerror(errno));
    return(1);
    }
  return(0);
  }

static int nbKitLink(char *home,char *caboodle,char *directory){
  char pathname[1024];
  int len,rc;
  DIR *dir;

  len=snprintf(pathname,sizeof(pathname),"%s/.nb",home);
  if(len<0 || len>=sizeof(pathname)){
    fprintf(stderr,"Unable to link %s - link path too long for buffer\n",caboodle);
    return(1);
    }
  dir=opendir(pathname);
  if(!dir){
    if(mkdir(pathname,S_IRWXU)<0){
      fprintf(stderr,"Unable to make directory %s - %s\n",pathname,strerror(errno));
      return(1);
      }
    }
  else closedir(dir);
  len=snprintf(pathname,sizeof(pathname),"%s/.nb/caboodle",home);
  if(len<0 || len>=sizeof(pathname)){
    fprintf(stderr,"Unable to link %s - link path too long for buffer\n",caboodle);
    return(1);
    }
  dir=opendir(pathname);
  if(!dir){
    if(mkdir(pathname,S_IRWXU)<0){
      fprintf(stderr,"Unable to make directory %s - %s\n",pathname,strerror(errno));
      return(1);
      }
    }
  else closedir(dir);
  len=snprintf(pathname,sizeof(pathname),"%s/.nb/caboodle/%s",home,caboodle);
  if(len<0 || len>=sizeof(pathname)){
    fprintf(stderr,"Unable to link %s - link path too long for buffer\n",caboodle);
    return(1);
    }
  rc=unlink(pathname);
  if(rc<0 && errno!=ENOENT){
    fprintf(stderr,"Unable to unlink %s - %s\n",pathname,strerror(errno));
    return(1);
    }
  rc=symlink(directory,pathname);
  if(rc<0){
    fprintf(stderr,"Unable to link %s to %s - %s\n",pathname,directory,strerror(errno));
    return(1);
    }
  return(0);
  }

static int nbKitCaboodle(char *home,char *caboodle,char *cabdir,size_t size){
  char linkname[1024];
  int len;
  ssize_t linksize;

  len=snprintf(linkname,sizeof(linkname),"%s/.nb/caboodle/%s",home,caboodle);
  if(len<0 || len>=size){
    fprintf(stderr,"Unable to lookup caboodle %s/.nb/caboodle/%s - path too long\n",home,caboodle);
    return(1);
    }
  linksize=readlink(linkname,cabdir,size);
  if(linksize<0){
    if(errno==ENOENT) fprintf(stderr,"Caboodle %s not defined\n",caboodle);
    else fprintf(stderr,"Error reading link %s - %s\n",linkname,strerror(errno));
    return(1);
    }
  if(linksize>=size){
    fprintf(stderr,"Error reading link %s - caboodle directory path too long for buffer\n",linkname);
    return(1);
    }
  *(cabdir+linksize)=0;
  return(0);
  }
static int nbKitDirectory(char *home,char *caboodle){
  char cabdir[1024];
  int rc;

  rc=nbKitCaboodle(home,caboodle,cabdir,sizeof(cabdir));
  if(rc) return(rc);
  printf("%s",cabdir);
  return(0);
  }

static int nbKitCmd(char *home,char *caboodle,int argc,char *argv[]){
  char pathname[1024];
  char cabdir[1024];
  char *tarcmd=cabdir;  // use same area for tar command if needed
  int len,rc;

  rc=nbKitCaboodle(home,caboodle,cabdir,sizeof(cabdir));
  if(rc) return(rc);
  len=snprintf(pathname,sizeof(pathname),"%s/bin/nbkit_",cabdir);
  if(len<0 || len>=sizeof(pathname)){
    fprintf(stderr,"Unable to execute %s/nbkit_ - path too long for buffer\n",cabdir);
    return(1);
    }
  rc=chdir(cabdir);
  if(rc<0){
    fprintf(stderr,"Unable to change working directory to %s - %s\n",cabdir,strerror(errno));
    return(1);
    }
  execvp(pathname,argv);
  if(errno==ENOENT){
    if(argc==4 && strcmp(argv[2],"use")==0 && strstr(argv[3],"caboodle")!=NULL){
      len=snprintf(tarcmd,sizeof(cabdir),"tar -xf %s bin/nbkit_",argv[3]);
      if(len<0 || len>=sizeof(cabdir)){
        fprintf(stderr,"Unable to invoke tar command - too large for buffer\n");
        return(1);
        }
      rc=system(tarcmd);
      if(rc==0){
        execvp(pathname,argv);
        }
      else{ 
        fprintf(stderr,"\nAttempt to initialize the caboodle with the following command failed rc=%d\n",rc);
        fprintf(stderr,"  %s\n",tarcmd);
        return(1);
        }
      }
    else{
      fprintf(stderr,"File '%s' not found\n\n",pathname);
      fprintf(stderr,"Issue the following command to initialize the caboodle:\n");
      fprintf(stderr,"  nbkit %s use <caboodle-kit-archive>\n\n",caboodle);
      fprintf(stderr,"To locate the caboodle kit archive file:\n");
      fprintf(stderr,"  nbkit -k | grep caboodle\n\n");
      fprintf(stderr,"If not found, install nbkit-caboodle package.\n");
      fprintf(stderr,"See documentation at http://nodebrain.org\n");
      return(1);
      }
    }
  fprintf(stderr,"Unable to invoke the following command - %s\n",strerror(errno));
  fprintf(stderr,"  %s\n",pathname);
  return(1);
  }

int nbKit(int argc,char *argv[]){
  struct passwd *pwd;
  
  pwd=getpwuid(getuid());
  if(!pwd){
    fprintf(stderr,"Unable to local user account information.\n");
    return(1);
    }

  // Handle:  nbkit
  if(argc<=1) return(nbKitUsage());

  // Handle:  nbkit --option
  if(*argv[1]=='-'){
    if(strcmp(argv[1],"-c")==0 || strcmp(argv[1],"--caboodles")==0){
      if(argc>2){
        printf("Unexpected parameters after -c or --caboodles option\n");
        return(1);
        }
      return(nbKitCaboodles(pwd->pw_dir));
      }
    if(strcmp(argv[1],"-k")==0 || strcmp(argv[1],"--kits")==0){
      if(argc>2){
        printf("Unexpected parameters after -k or --kits option\n");
        return(1);
        }
      return(nbKitKits());
      }
    return(nbKitUsage());
    }

  // Handle:  nbkit <caboodle>
  if(argc==2) return(nbKitDirectory(pwd->pw_dir,argv[1]));

  // Handle:  nbkit <caboodle> <verb> ... 
  // Note: argc>2
  if(strcmp(argv[2],"link")==0){
    if(argc>4){
      printf("Link command passed too many parameters - only three allowed\n");
      printf("nbkit <caboodle> link [<directory>]\n");
      return(1);
      }
    if(argc==0){
      printf("Link command does not support optional directory - three parameters required\n");
      printf("nbkit <caboodle> link <directory>\n");
      return(1);
      }
    return(nbKitLink(pwd->pw_dir,argv[1],argv[3]));
    }
  if(strcmp(argv[2],"unlink")==0){
    if(argc>3){
      printf("Unlink command passed too many parameters - only two allowed\n");
      printf("nbkit <caboodle> unlink\n");
      return(1);
      }
    return(nbKitUnlink(pwd->pw_dir,argv[1]));
    }
  return(nbKitCmd(pwd->pw_dir,argv[1],argc,argv));
  }
